src/txn.c

```diff
@@ -317,13 +337,14 @@ txnDecoderState_e txn_next_elem(txn_state_t *txn) {
 
 void txn_init(txn_state_t *txn, uint16_t sigIndex, uint32_t changeIndex) {
     memset(txn, 0, sizeof(txn_state_t));
-    txn->buflen = txn->pos = txn->sliceIndex = txn->displayIndex = txn->sliceLen = txn->valLen = 0;
-    txn->elemType = -1;  // first increment brings it to SC_INPUT
     txn->sigIndex = sigIndex;
 
+    txn->elementIndex = 0;
+    txn->elements[txn->elementIndex].elemType = -1;  // first increment brings it to SC_INPUT
+
     cx_ecfp_public_key_t publicKey = {0};
-    deriveSiaKeypair(changeIndex, NULL, &publicKey);
-    pubkeyToSiaAddress(&txn->changeAddr, &publicKey);
+    deriveSiaPublicKey(changeIndex, publicKey.W);
+    pubkeyToSiaAddress((char *)&txn->changeAddr, &publicKey);
```

Comment:
Same you should be able to allocate a simble uint8_t buffer instead of a cx_ecfp_public_key_t